File: Day7/find the largest subarray whose sum is less than k .cpp
#include<bits/stdc++.h>
using namespace std;

int atMostSum(vector<int> arr, int k){
    int n=arr.size();
    int count=0;
   
    //Brute Force Approach
    //TC=0(n2) and SC=0(1)
    // for(int i=0;i<n;i++){
    //      int sum=0;
    //     for(int j=i;j<n;j++){
    //         sum+=arr[j];
    //         if(sum<=k){
    //             count=max(count,j-i+1);
    //         }
    //     }
    // }
    // return count;


    //optimized approach using two pointer if all the intergers are positive

    int i=0;
    int sum=0;
    for(int j=0;j<n;j++){
        sum+=arr[j];
        while(sum>k && i<=j){
            sum-=arr[i];
            i++;
        }
        count=max(count,j-i+1);
    }
    return count;
}





int main() {
    vector<int> arr = { 1, 2, 1, 0, 1, 1, 0 };
    int k = 4;
    int res = atMostSum(arr, k);
    cout << res << endl;
    return 0;
}

File: Day15_Revision/find the largest subarray whose sum is less than k.py
def largestsubarray(arr,k):
    n=len(arr)
    max_len=0
    sum=0
    i=0

    #Brute Force Solution TC=0(n2) and SC=01(1)

    # for i in range(0,n):
    #    sum=0
    #     for j in range(i,n):
    #         sum+=arr[j]

    #         if(sum<k):
    #             max_len=max(max_len,j-i+1)


    #Optimized Approach using two pointer
    for j in range(0,n):
        sum+=arr[j]

        while sum>=k and i<=j:
            sum-=arr[i]
            i+=1

        max_len=max(max_len,j-i+1)

    return max_len











